main/lsp/protocol.cc

```diff
@@ -171,36 +237,42 @@ unique_ptr<core::GlobalState> LSPLoop::runLSP() {
             }
         });
 
+    // Bridges the gap between the {reader, watchman} threads and the coordinator thread.
+    auto preprocessingThread =
+        startPreprocessorThread(incomingQueue, incomingMtx, processingQueue, processingMtx, logger);
+
     mainThreadId = this_thread::get_id();
     unique_ptr<core::GlobalState> gs;
     {
         // Ensure Watchman thread gets unstuck when thread exits.
         NotifyNotificationOnDestruction notify(initializedNotification);
+        bool exitProcessed = false;
         while (true) {
             unique_ptr<LSPMessage> msg;
             bool hasMoreMessages;
             {
-                absl::MutexLock lck(&mtx);
+                absl::MutexLock lck(&processingMtx);
                 Timer timeit(logger, "idle");
-                mtx.Await(absl::Condition(
-                    +[](LSPLoop::QueueState *guardedState) -> bool {
-                        return guardedState->terminate ||
-                               (!guardedState->paused && !guardedState->pendingRequests.empty());
+                processingMtx.Await(absl::Condition(
+                    +[](LSPLoop::QueueState *processingQueue) -> bool {
+                        return processingQueue->terminate ||
+                               (!processingQueue->paused && !processingQueue->pendingRequests.empty());
                     },
-                    &guardedState));
-                ENFORCE(!guardedState.paused);
-                if (guardedState.terminate) {
-                    if (guardedState.errorCode != 0) {
+                    &processingQueue));
+                ENFORCE(!processingQueue.paused);
+                if (processingQueue.terminate) {
+                    if (processingQueue.errorCode != 0) {
                         // Abnormal termination.
-                        throw options::EarlyReturnWithCode(guardedState.errorCode);
-                    } else if (guardedState.pendingRequests.empty()) {
-                        // Normal termination. Wait until all pending requests finish.
+                        throw options::EarlyReturnWithCode(processingQueue.errorCode);
+                    } else if (exitProcessed || processingQueue.pendingRequests.empty()) {
```

Comment:
The only new logic here is `exitProcessed`, which prevents the queue from processing extra messages that occur after an `exit` request.

An existing test was supposed to test this edge case, but due to the single queue mutex it never actually exercised it. Once I added the second mutex, letting the other threads enqueue messages while this thread pulled from the queue, the test started failing.

Now, the thread exits after it processes an `exit` request, as it should. It ignores any extra messages in the queue.